test/cttestsrv/server.go

```diff
@@ -0,0 +1,289 @@
+package cttestsrv
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/certificate-transparency-go"
+)
+
+// Personality describes the configuration & behaviour of a test CT
+// IntegrationSrv
+type Personality struct {
+	// Port (and optionally IP) to listen on
+	Addr string
+	// Private key for signing SCTs
+	// Generate your own with:
+	// openssl ecparam -name prime256v1 -genkey -outform der -noout | base64 -w 0
+	PrivKey string
+	// If present, sleep for the given number of seconds before replying. Each
+	// request uses the next number in the list, eventually cycling through.
+	LatencySchedule []float64
+}
+
+// IntegrationSrv is an instance of a CT test server.
+type IntegrationSrv struct {
+	sync.RWMutex
+	logger *log.Logger
+
+	// key is the log's private key used to sign STHs and SCTs
+	key *ecdsa.PrivateKey
+
+	// PubKey is the log's public key in base64 encoded format (e.g. the "log ID")
+	PubKey string
+
+	// latencySchedule holds the latency schedule from the Personality that was
+	// used to create the IntegrationSrv instance. This controls the latency
+	// pattern of add-chain/add-pre-chain responses.
+	latencySchedule []float64
+	// latencyItem is the latency used for the next add-chain/add-pre-chain and is
```

Comment:
need to fix this comment: its not the latency used, its the index into latencySchedule.